gpio/src/gpio.rs

```diff
@@ -0,0 +1,943 @@
+
+/// Virtio GPIO Configuration
+#[derive(Clone, Debug, PartialEq)]
+#[repr(C)]
+pub(crate) struct VirtioGpioConfig {
+    pub(crate) ngpio: Le16,
+    pub(crate) padding: Le16,
+    pub(crate) gpio_names_size: Le32,
+}
+
+/// Trait that represents an GPIO Device.
+///
+/// This trait is introduced for development purposes only, and should not
+/// be used outside of this crate. The purpose of this trait is to provide a
+/// mock implementation for the GPIO driver so that we can test the GPIO
+/// functionality without the need of a physical device.
+pub(crate) trait GpioDevice: Send + Sync + 'static {
+    fn open(device: u32) -> Result<Self>
+    where
+        Self: Sized;
+
+    fn get_num_lines(&self) -> Result<u16>;
+    fn get_line_name(&self, gpio: u16) -> Result<String>;
+    fn get_direction(&self, gpio: u16) -> Result<u8>;
+    fn set_direction(&self, gpio: u16, dir: u8, value: u32) -> Result<()>;
+    fn get_value(&self, gpio: u16) -> Result<u8>;
+    fn set_value(&self, gpio: u16, value: u32) -> Result<()>;
+
+    fn set_irq_type(&self, gpio: u16, value: u16) -> Result<()>;
+    fn wait_for_interrupt(&self, gpio: u16) -> Result<()>;
+}
+
+pub(crate) struct PhysDeviceState {
+    // See wait_for_interrupt() for explanation of Arc.
+    request: Option<Arc<LineRequest>>,
+    buffer: Option<EdgeEventBuffer>,
+}
+
+pub(crate) struct PhysDevice {
+    chip: Chip,
+    ngpio: u16,
+    state: Vec<RwLock<PhysDeviceState>>,
+}
+
+unsafe impl Send for PhysDevice {}
+unsafe impl Sync for PhysDevice {}
+
+impl GpioDevice for PhysDevice {
+    fn open(device: u32) -> Result<Self>
+    where
+        Self: Sized,
+    {
+        let path = format!("/dev/gpiochip{}", device);
+        let chip = Chip::open(&path).map_err(Error::GpiodFailed)?;
+        let ngpio = chip.get_num_lines() as u16;
+
+        // Can't set a vector to all None easily
+        let mut state: Vec<RwLock<PhysDeviceState>> = Vec::new();
+        state.resize_with(ngpio as usize, || {
+            RwLock::new(PhysDeviceState {
+                request: None,
+                buffer: None,
+            })
+        });
+
+        Ok(PhysDevice { chip, ngpio, state })
+    }
+
+    fn get_num_lines(&self) -> Result<u16> {
+        Ok(self.ngpio)
+    }
+
+    fn get_line_name(&self, gpio: u16) -> Result<String> {
+        let line_info = self
+            .chip
+            .line_info(gpio.into())
+            .map_err(Error::GpiodFailed)?;
+
+        Ok(line_info.get_name().unwrap_or("").to_string())
+    }
+
+    fn get_direction(&self, gpio: u16) -> Result<u8> {
+        let line_info = self
+            .chip
+            .line_info(gpio.into())
+            .map_err(Error::GpiodFailed)?;
+
+        Ok(
+            match line_info.get_direction().map_err(Error::GpiodFailed)? {
+                Direction::AsIs => VIRTIO_GPIO_DIRECTION_NONE,
+                Direction::Input => VIRTIO_GPIO_DIRECTION_IN,
+                Direction::Output => VIRTIO_GPIO_DIRECTION_OUT,
```

Comment:
Not sure why we have to re-define enums from the library into some of our own.

Reply:
Library (gpiolib) is just one of the implementations of the device, we can have more, like a bare-metal one. And so need to have our own values (as defined by virtio-spec).

Reply:
They are already declared under the following comment:

/// Virtio specification definitions

I thought that would be enough ?

Reply:
I took that to mean "definitions related to virtio" rather than "part of the virtio specification".